generic3g/actions/AccumulatorAction.F90

```diff
@@ -0,0 +1,302 @@
+      type(ESMF_Field), allocatable :: accumulation_field
+      type(ESMF_Field), allocatable :: counter_field
+   contains
+      procedure :: initialize => initialize_action
+      procedure :: update => update_action
+      procedure :: invalidate => invalidate_action
+      procedure, private :: has_coupler
+      procedure, private :: has_counter
+      procedure, private :: is_active
+   end type AccumulatorAction
+
+   interface AccumulatorAction
+      module procedure :: construct_accumulator_action
+   end interface AccumulatorAction
+
+   enum, bind(c)
+      enumerator :: SIMPLE_ACCUMULATOR
+      enumerator :: MEAN_ACCUMULATOR
+      enumerator :: MAX_ACCUMULATOR
+      enumerator :: MIN_ACCUMULATOR
+   end enum
+
+contains
+
+   function construct_accumulator_action(accumulator_type)
+      type(AccumulatorAction) :: acc
+      integer(kind=SIMPLE_ACCUMULATOR), intent(in) :: accumulator_type
+
+      acc%accumulator_type = accumulator_type
+
+   end function construct_accumulator_action
+
+   subroutine set_accumulator_action(acc) result(acc)
+      class(AccumulatorAction), intent(inout) :: acc
+      procedure(UpdateField), pointer :: accumulate => null()
+      procedure(UpdateField), pointer :: increment => null()
+      procedure(ModifyField), pointer :: clear => null()
+      procedure(UpdateField), pointer :: couple => null()
+
+      clear => clear_to_zero
+      select case(acc%accumulator_type)
+         case MEAN_ACCUMULATOR
+            accumulate => accumulate_add
+            couple => couple_mean
+            increment => increment_counter
+         case MIN_ACCUMULATOR
+            accumulate => accumulate_min
+            clear => clear_to_undef
+         case MAX_ACCUMULATOR
+            accumulate => accumulate_max
+            clear => clear_to_undef
+         case default
+            _FAIL("Unsupported AccumulatorAction")
+      end select
+      acc = AccumulatorAction(accumulate, increment, couple, clear)
+
+   end subroutine set_accumulator_action
+
+   subroutine initialize_action(this, importState, exportState, clock, rc)
+      class(AccumulatorAction), intent(inout) :: this
+      type(ESMF_State) :: importState
+      type(ESMF_State) :: exportState
+      type(ESMF_Clock) :: clock
+      integer, optional, intent(out) :: rc
+      type(ESMF_Field) :: import_field, export_field, field
+
+      call this%set_accumulator_action(_RC)
+      _ASSERT(assigned(this%couple) .eqv. assigned(this%increment), 'couple requires increment')
+      call get_field(importState, import_field, _RC)
+      call get_field(exportState, export_field, _RC)
+      call FieldCopy(import_field, field, _RC)
+      call acc%clear(field, _RC)
+      acc%accumulation_field = field
+      if(acc%has_coupler()) then
+         call FieldCopy(import_field, field, _RC)
+         call FieldReallocate(field, typekind=COUNTER_TYPEKIND, _RC)
+         call this%clear(field, _RC)
+         acc%counter_field = field
+      end if
```

Comment:
Is this just a way to create a field with the same structure but with 0's in the values?

Reply:
Yes. Is there a simpler way?